keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1500 @@
+
+### Feature gate
+
+Literal feature gates are not super useful for CLI features, so SIG CLI has a
+similar mechanism to gate features via environment variables.  We will use a
+variable named `"KUBECTL_KYAML"` to gate this feature.
+
+The output flags for JSON and YAML are handled in
+`k8s.io/cli-runtime/pkg/genericclioptions`.  Unfortunately, that package is
+already transitively depended on by `k8s.io/kubectl/pkg/cmd/util` where the CLI
+gates are defined. Rather than trying to break the import cycle, we will just
+check the environment variable directly, without the utils package.
+
+### Future compatibility
+
+KYAML is, and will remain, a strict subset of YAML.  Compliant YAML parsers
+should always be able to parse KYAML data and we will endeavour to make sure
+they can all retain fidelity with comments.
+
+It may be that, over time, we find some aspects of KYAML's exact rendering that
+can be inmproved, and we will consider those opportunities. The implication of
+this is that clients which rely on the byte-exact results of KYAML's rendering
+might be broken by future changes. We will try to avoid this, but we will not
+guarantee it.
+
+### Packaging
+
+We will create a new `kyaml` package under `sigs.k8s.io/yaml` for the rendering
+logic and put the printer under `staging/src/k8s.io/cli-runtime/pkg/printers`,
+using the renderer.
+
+### Components
+
+The primary component affected by this KEP is `kubectl`, but any other
+component which uses `k8s.io/cli-runtime/pkg/printers` and
+`k8s.io/cli-runtime/pkg/genericclioptions` will get KYAML support "for free".
+
+### Tooling
+
+k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and
+ensures they are formatted “correctly”. The only meaning of “correct” is “block
+style”, and it corrupts KYAML input today (aside: it also does not handle
+multi-documents).  If we intend to convert examples, we need to fix this.
+
+We propose to add flags to it which specify to use “conventional” YAML or KYAML
+output, and set the default to KYAML (or to change the parent script to always
+specify KYAML).
+
+This tooling will preserve comments as best it can. Unfortunately, go-yaml (our
+YAML library) [does not always handle comments properly](https://github.com/yaml/go-yaml/issues/8).
+It is possible that some comments will be formatted wrongly, or lost entirely.
+The affected comments are not representable in YAML block-style, so this is not
+exactly a regression.
+
+A guiding principle for the tooling is that whatever output the tool produces,
+when fed back into the same version of the tool, should not be modified again.
+As mentioned above, we MIGHT make formatting changes over time, so this
+property may not hold across versions of the tooling.
+
+Complex YAML input, with things like anchors, aliases, explicit tags (!!...),
+and global tags (%TAG !...) may be simplified when rendered to KYAML, as long
+as the resulting unmarshaled object would be identical.
```

Comment:
I'm not quite sure what you mean "maybe be simplified" here. Care to explain?

Reply:
YAML allows things like aliases which will be dereferenced when run through this:

e.g.

```
$ go run ./yaml_renderer.go 
- &flag Apple
- Beachball
- Cartoon
- Duckface
- *flag
---
[
  "Apple",
  "Beachball",
  "Cartoon",
  "Duckface",
  "Apple",
]
```

Reply:
Added note